File: LeetCode/2000-minimum-speed-to-arrive-on-time/minimum-speed-to-arrive-on-time.java
class Solution {
    public int minSpeedOnTime(int[] dist, double hour) {
        int min = 0, max = (int) Math.pow(10, 7), sol = -1;
        
        while (min <= max) {
            int mid = (min + max)/2;
            if (check(dist, hour, mid)) {
                sol = mid;
                max = mid - 1;
            } else {
                min = mid + 1;
            } //if
        } //while
        
        return sol;
    } //minSeped

    boolean check (int [] dists, double deadline, int speed) {
        double currTime = 0.0;

        for (int i = 0; i < dists.length; i++) {
            currTime += ((dists[i] * 1.0)/speed);
            if (i != dists.length - 1) currTime = Math.ceil(currTime);
        } //for

        return currTime <= deadline;
    } //can
}

File: LeetCode/2000-minimum-speed-to-arrive-on-time/minimum-speed-to-arrive-on-time.cpp
class Solution {
public:
    int minSpeedOnTime(vector<int>& dists, double deadline) {
        if (dists.size() - 1 >= deadline) return -1;
        int l = 0, n = dists.size(), r = 1e7;

        while (l <= r) {
            int m = l + (r - l)/2;
            bool can = check(dists, deadline, m);
            if (can) {
                r = --m;
            } else {
                l = ++m;
            } //if
        } //while   

        return l == 0 ? 1 : l;
    } //minSpeedOnTime

    bool check (vector<int>& dists, double deadline, int speed) {
        double currTime = 0.0;

        for (int i = 0; i < dists.size(); i++) {
            currTime += ((dists[i] * 1.0)/speed);
            if (i != dists.size() - 1) currTime = ceil(currTime);
        } //for

        return currTime <= deadline;
    } //can
};